### PR TITLE
Implement more Either

### PR DESCRIPTION
**Pull request by stephencelis**
Still a bit of a WIP (tests, even more `Either`), but wanted to push what I had so far.

### CODE REVIEW

File: Sources/Prelude/Either.swift [hunk 1]
```diff
@@ -17,12 +17,81 @@ extension Either {
...
+}
+
+public func >>- <A, B, C>(b2c: (B) -> Either<A, C>, b: Either<A, B>) -> Either<A, C> {
+  return b.flatMap(b2c)
+}
+
+// MARK: - Semigroup
+
+public func <> <L, R: Semigroup>(lhs: Either<L, R>, rhs: Either<L, R>) -> Either<L, R> {
+  return lhs.flatMap { lhs in rhs.map { rhs in lhs <> rhs } }
```
Comment (stephencelis) [inline Sources/Prelude/Either.swift]: For whatever reason, when I tried the following:

``` swift
return (<>) <¢> lhs <*> rhs
// or
return pure(<>) <*> lhs <*> rhs
```

I got the following error:

> Ambiguous use of operator '<>'

Haven't dug deep enough to figure out if it's a bug in the compiler or if we can disambiguate.

File: Sources/Prelude/Monoid.swift [hunk 2]
```diff
@@ -2,8 +2,8 @@ public protocol Monoid: Semigroup {
   static var e: Self { get }
 }
 
-public func concat<M: Monoid>(_ xs: [M]) -> M {
-  return xs.reduce(M.e, <>)
+public func concat<S: Sequence>(_ xs: S) -> S.Element where S.Element: Monoid {
```
Comment (stephencelis) [inline Sources/Prelude/Monoid.swift]: Just some more `Array` to `Sequence` generalization.

File: Sources/Prelude/Either.swift (hunk 1 above)
Comment (mbrandonw) [inline Sources/Prelude/Either.swift]: Doesn't it need to be curried?

Comment (stephencelis) [inline Sources/Prelude/Either.swift]: Lol! Yes! Been living in curry land recently is breaking my brain.

File: Sources/Prelude/Function.swift [hunk 3]
```diff
@@ -29,3 +29,11 @@ public func flip<A, B, C>(_ f: @escaping (A) -> (B) -> C) -> (B) -> (A) -> C {
     }
   }
 }
+
+public func curry<A, B, C>(_ f: @escaping (A, B) -> C) -> (A) -> (B) -> C {
```
Comment (stephencelis) [inline Sources/Prelude/Function.swift]: We might wanna consider depending on https://github.com/thoughtbot/Curry in the future, or we could start by having a smaller set of overloads here in Prelude.

**Pull request merged**